### PR TITLE
RavenDB-6812 Fixing an issue where dispose might be running while ref…

### PR DESCRIPTION
**Pull request by talweiss1982**
…reshing ambassador is also running and that may cause unexpected enumeration modification exception

### CODE REVIEW

File: src/Raven.Server/Rachis/Leader.cs [hunk 1]
```diff
@@ -557,18 +557,18 @@ public void Dispose()
                 _thread.Join();
 
             var ae = new ExceptionAggregator("Could not properly dispose Leader");
```
Comment (ayende) [inline src/Raven.Server/Rachis/Leader.cs]: This didn't actually fix the problem. We are still exposed to the modification during the `ToArray`

**Comment by talweiss1982** [comment]
not sure what to do if we can't take the lock during dispose...
this should not take long to refresh the ambassadors worst case we wait abit

File: src/Raven.Server/Rachis/Leader.cs [hunk 2]
```diff
@@ -104,94 +104,111 @@ public void StepDown()
...
-            Task.Run(() =>
+            finally
             {
-                foreach (var ambasaddor in old)
-                {
-                    // it is not used by anything else, so we can close it
-                    ambasaddor.Value.Dispose();
-                }
-            });
+                Monitor.Exit(topologyLocker);
```
Comment (ppekrol) [inline src/Raven.Server/Rachis/Leader.cs]: if (lockTaken) Monitor.Exit(topologyLocker) ?

File: src/Raven.Server/Rachis/Leader.cs [hunk 3]
```diff
@@ -529,57 +546,80 @@ public Task<long> PutAsync(BlittableJsonReaderObject cmd)
...
+                _noop.Dispose();
+                if (_engine.Log.IsInfoEnabled)
+                {
+                    _engine.Log.Info($"Leader {_engine.Tag}: Dispose");
+                }
+            }
+            finally
             {
-                _engine.Log.Info($"Leader {_engine.Tag}: Dispose");
+                Monitor.Exit(topologyLocker);
```
Comment (ppekrol) [inline src/Raven.Server/Rachis/Leader.cs]: if (lockTaken) Monitor.Exit(topologyLocker) ?

File: src/Raven.Server/Rachis/Leader.cs [hunk 4]
```diff
@@ -104,94 +104,111 @@ public void StepDown()
...
+            try
             {
-                foreach (var peer in peers)
+                //This only means we are been disposed so we can quit now
+                if (lockTaken == false)
                 {
-                    old[peer.Key] = peer.Value;
+                    if (_engine.Log.IsInfoEnabled)
+                    {
+                        _engine.Log.Info($"Leader {_engine.Tag}: Skipping refreshing ambassadors because we are been disposed of");
```
Comment (ayende) [inline src/Raven.Server/Rachis/Leader.cs:116]: `been` -> `being`

File: src/Raven.Server/Rachis/Leader.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Rachis/Leader.cs]: Yes

File: src/Raven.Server/Rachis/Leader.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Rachis/Leader.cs]: Yes

**Review by ayende — CHANGES_REQUESTED** [review]

**Pull request merged**